x-pack/plugins/apm/public/context/url_params_context/helpers.ts

```diff
@@ -30,24 +51,38 @@ export function getDateRange({
 }) {
   // If the previous state had the same range, just return that instead of calculating a new range.
   if (state.rangeFrom === rangeFrom && state.rangeTo === rangeTo) {
-    return { start: state.start, end: state.end };
+    return {
+      start: state.start,
+      end: state.end,
+      exactStart: state.exactStart,
+      exactEnd: state.exactEnd,
+    };
   }
-
   const start = getParsedDate(rangeFrom);
   const end = getParsedDate(rangeTo, { roundUp: true });
 
   // `getParsedDate` will return undefined for invalid or empty dates. We return
   // the previous state if either date is undefined.
   if (!start || !end) {
-    return { start: state.start, end: state.end };
+    return {
+      start: state.start,
+      exactStart: state.start,
+      end: state.end,
+      exactEnd: state.end,
+    };
   }
 
   // rounds down start to minute
   const roundedStart = moment(start).startOf('minute');
 
   return {
     start: roundedStart.toISOString(),
+    exactStart:
+      getExactDate(rangeFrom)?.toISOString() || roundedStart.toISOString(),
     end: end.toISOString(),
+    exactEnd:
+      getExactDate(rangeTo, { roundUp: true })?.toISOString() ||
+      end.toISOString(),
   };
```

Comment:
You could break it up like:

```suggestion
const roundedStart = moment(start).startOf("minute"); // rounds down to minute
const exactStart = getExactDate(rangeFrom) || roundedStart;
const exactEnd = getExactDate(rangeTo, { roundUp: true }) || end;

return {
  start: roundedStart.toISOString(),
  exactStart: exactStart.toISOString(),
  end: end.toISOString(),
  exactEnd: exactEnd.toISOString(),
};

```